### PR TITLE
[libc++][hardening] Add hardening assertions to std::deque

### LINKED ISSUE #63809: deque is missing safe libc++ hardening checks
https://libcxx.llvm.org/UsingLibcxx.html#enabling-the-safe-libc-mode

vector's `operator[]` uses `_LIBCPP_ASSERT_UNCATEGORIZED`: https://github.com/llvm/llvm-project/blob/7b64d761098fdbb426b27ca58a9dc8f31917dcba/libcxx/include/vector#L1534C5-L1534C33

deque's `operator[]` just reads OOB without an assert: https://github.com/llvm/llvm-project/blob/7b64d761098fdbb426b27ca58a9dc8f31917dcba/libcxx/include/deque#L1645-L1661

Same for [`deque::front()`](https://github.com/llvm/llvm-project/blob/7b64d761098fdbb426b27ca58a9dc8f31917dcba/libcxx/include/deque#L1699) and other methods that return a reference.

### PR DESCRIPTION
**Pull request by davidben**
This aligns std::deque with std::vector w.r.t. hardening checks. There's probably more that can be done with iterators, but start with this.

This caught a bug with one of libc++'s tests. One of the erase calls in asan_caterpillar.pass.cpp was a no-op because the iterators were in the other order. (deque::erase happened to cleanly do nothing when the distance is negative.)

Fixes #63809

### CODE REVIEW

File: libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// <deque>
+
+// Test hardening assertions for std::deque.
+
+// REQUIRES: has-unix-headers
```
Comment (var-const) [inline libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp:13]: Nit: for completeness' sake, this also needs:
```
// UNSUPPORTED: c++03
// XFAIL: libcpp-hardening-mode=debug && availability-verbose_abort-missing
```
(due to the way `check_assertion.h` is implemented). I plan to add a new feature like `has-check-assertion` to wrap all those checks together, but for now it all has to be marked explicitly in tests relying on `check_assertion.h`.

**Review by var-const — APPROVED** [review]
LGTM, thanks a lot for working on this!

File: libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp (hunk 1 above)
Comment (davidben) [reply libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp:13]: Done

File: libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    TEST_LIBCPP_ASSERT_FAILURE(cc[0], "deque::operator[] index out of bounds");
+  }
+
+  c.push_back(1);
+  c.push_back(2);
+  c.push_back(3);
+  TEST_LIBCPP_ASSERT_FAILURE(c[3], "deque::operator[] index out of bounds");
+  TEST_LIBCPP_ASSERT_FAILURE(c[100], "deque::operator[] index out of bounds");
+
+  // Repeat the test with a const reference to test the const overloads.
```
Comment (danakj) [inline libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp:44]: The test cases above are also on a const reference, aren't they? Are we missing mutable overloads then?

Comment (danakj) [reply libcxx/test/libcxx/containers/sequences/deque/assert.pass.cpp:44]: Oh nevermind, sorry, I was looking at the front() ones but this is [].

**Pull request merged**